### PR TITLE
3 Feature backend storage of user data

### PR DESCRIPTION
**Pull request by EricKarpovits**
The backend storage has been completed. Any object of any type can be stored given a unique string key. 
- I created the gateway interface for the Database 
- Created the Database
- Created tests for the Database

Key files to look over
- src/main/java/database/Database.java
- src/main/java/database/DatabaseGateway.java

Let me know if there is anything I need to fix.

### CODE REVIEW

File: src/main/java/entities/User.java [hunk 1]
```diff
@@ -1,7 +1,7 @@
-package entities;
-
-public class User extends AbstractUser{
```
Comment (fishyvishy) [inline src/main/java/entities/User.java:3]: It says that AbstractUser has been deleted, so User can't inherit from AbstractUser. Is this a mistake on my page?

Comment (EricKarpovits) [reply src/main/java/entities/User.java:3]: Talk to @Shaffaan about it but apparently, we no longer use the AbstractUser class. Let me know if I should be keeping it though.

File: src/main/java/fridge/Fridge.java [hunk 2]
```diff
@@ -1,37 +1,63 @@
...
+        return str;
+    }
+
     /**
      * Print the list of ingredients in Ingredient
      * */
+
     @Override
     public String toString() {
         return super.toString();
```
Comment (fishyvishy) [inline src/main/java/fridge/Fridge.java:67]: I don't understand why this method is necessary if it calls on the super.

Comment (EricKarpovits) [reply src/main/java/fridge/Fridge.java:67]: This isn't something I did (green is what I did). So I'm not too sure myself why it is necessary to call super but I'm pretty sure @Shaffaan changed the Fridge class too.

File: src/main/java/database/Database.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    /**
+     * Gets the first and only key in the Database which is the username.
+     *
+     * @return a String that is the username if it was stored otherwise an empty String.
+     */
+    @Override
+    public String getUsername() {
+        if (!database.isEmpty()) {
+            Object username = this.database.keySet().iterator().next();
```
Comment (rickychen175) [inline src/main/java/database/Database.java:112]: While this doesn't change function behavior, I think using database and this.database together could be confusing in the future.

**Review by rickychen175 — APPROVED** [review]
The database looks good.

File: src/main/java/database/Database.java (hunk 3 above)
Comment (EricKarpovits) [reply src/main/java/database/Database.java:112]: Sure I'll fix it to use this.database to be consistent. Thank you.

File: src/main/java/entities/User.java (hunk 1 above)
Comment (Shaffaan) [reply src/main/java/entities/User.java:3]: @fishyvishy I changed abstract user, and we do not have it anymore. The new classes are going to be a user interface and a common user which implements user interface. I will be sending a pull req for those really soon.

File: src/main/java/database/Database.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
+package database;
+
+import java.io.*;
+import java.util.HashMap;
+
```
Comment (fishyvishy) [inline src/main/java/database/Database.java:5]: A javadoc describing the database might be necessary.

**Review by Shaffaan — APPROVED** [review]
Great work eric and good job on the testing as well!

File: src/test/java/DatabaseTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    private DatabaseGateway database;
+    private Fridge fridge;
+
+    /**
+     * Setup method before each test
+     */
+    @BeforeEach
+    public void setUp() {
+        fridge = new Fridge();
+        Ingredient t = new Ingredient();
```
Comment (fishyvishy) [inline src/test/java/DatabaseTest.java:28]: Probably more a question for @Shaffaan. Is this the implementation that's going to be used for Ingredient. Would it be better to have the name passed in as argument?

**Review by fishyvishy — APPROVED** [review]
Database seems good. Add docstring to beginning of database class. Tests might have to be changed if Ingredient implementation changes.

File: src/main/java/database/Database.java (hunk 4 above)
Comment (EricKarpovits) [reply src/main/java/database/Database.java:5]: Sure I'll add it.

**Comment by EricKarpovits** [comment]
> Database seems good. Add docstring to the beginning of the database class. Tests might have to be changed if Ingredient implementation changes.

Yes the tests will have to be changed when the Ingredient implementation changes. The problem is that when you deserialize an object it returns an instance of the same object with a different code per se. So it is hard to test it. 
I used option 1 to test but I'm thinking to use option 2. 
https://www.clear.rice.edu/comp310/JavaResources/RMI/compare.htmlwhen you deserialize an object it

File: src/test/java/DatabaseTest.java (hunk 5 above)
Comment (EricKarpovits) [reply src/test/java/DatabaseTest.java:28]: Yes I think Shaffaan has fixed this.

**Pull request merged**